### ISSUE TITLE
False positives with urlDecodeUni

### ISSUE THREAD

**Issue by theseion** [coreruleset/coreruleset#4059]
> Hi.
> After updating the rules to version 4.12.0.
> I got a lot of false positives. And selectively. from some browsers the same url worked without blocks. And other browsers had problems and triggered 949110 
> And when opening the root page of my resource.
> after I returned htmlEntityDecode and removed urlDecodeUni in the rules.
> 921140
> 921150
> 921160
> 921151
> The problem disappeared.
> Where could be the bug and how to find it. Is it in urlDecodeUni or something in my resource code?
> 
> here is an example of a coded url for google ads 
> which some of them pass and some of them fail the rule.
> %7C - |
> %7B - {
> %7D - }
> utm_source=google&utm_medium=display&utm_campaign=kms%7C%7Bunix%7D%7Cgoogle%7Cazp&utm_content=729289294201&gad_source=5&gclid=EAIaIQob 

 _Originally posted by @vagrantsd in [#3824](https://github.com/coreruleset/coreruleset/issues/3824#issuecomment-2760358793)_

**Comment by theseion** [comment]
Hi @vagrantsd. I've tested the query string you've mentioned and it doesn't trigger any of the rules you've listed. It's impossible to tell what the issue is here, but I find it highly suspicious that you say that the browser seems to matter. That is very unlikely.

1. The rules that you've listed all test for injection attempts using line breaks, e.g., header injection. The query string might not be the culprit here
1. Please post an example that actually fails (complete request with `curl`, not just the URL)
2. You can test payloads yourself using the sandbox: https://coreruleset.org/docs/6-development/6-4-using-the-crs-sandbox
3. If you have access to the WAF log output you should see all the rules that trigger and the specific match that was found. If you don't know how to read it, you can post it here

If you can post the relevant lines of the log output we should be able to tell quickly what's going on.

**Comment by vagrantsd** [comment]
The problem is observed only on certain user agents. When navigating to the same resource. And nothing special is transferred. The test passed on visiting resources several thousand unique visits. And the problem is only on the given user agent:

Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/135.0.0.0 Safari/537.36
Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/135.0.0.0 Safari/537.36 Edg/135.0.0.0
Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/132.0.0.0 YaBrowser/25.2.0.0 Safari/537.36
If you go back to the settings before the update.
After I returned htmlEntityDecode and removed urlDecodeUni in the rules.
921140
921150
921160
921151
Then rule 949110 is no longer triggered for user agent data.

**Comment by fzipi** [comment]
Well... `%0a` is a linefeeed. You probably don't want to accept that.

**Comment by theseion** [comment]
What @fzipi said. Line feeds (`\n`) and carriage returns (`\n`) are blocked because attackers can abuse them to perform request splitting.

We have a dedicated [plugin](https://github.com/coreruleset/cpanel-rule-exclusions-plugin) for CPanel, which currently does not address this issue. Are you hosting CPanel yourself? We could update the plugin (if we find that it makes sense) and if you used it, your problem would be solved.

The alternative is for you to create an exclusion rule for this particular problem. Documentation on how to do that can be found [here](https://coreruleset.org/docs/2-how-crs-works/2-3-false-positives-and-tuning/#rule-exclusions).

**Comment by vagrantsd** [comment]
SecRuleRemoveById 921140
Unfortunately the plugin does not solve the problem when characters are passed in the cookie from cpanel.
The only solution is to disable the rule. 
But is it possible to disable it more precisely? Because here we are talking about the cookie and the rule itself does not check the cookie, but only the body of the request !

**Comment by theseion** [comment]
Yes. As described in the [docs](https://coreruleset.org/docs/2-how-crs-works/2-3-false-positives-and-tuning/#rule-exclusions), you would add a rule that dynamically disables rule `921140` using `ctl:ruleRemoveById`. The new rule needs to be loaded _before_ the rule you want to disable. For example:
```
SecRule REQUEST_COOKIES:/mixpanel$/ "@rx (?i)%0a" \
  "id:1000000,\
  phase:1,\
  t:none,\
  ctl:ruleRemoveById=921140"
```
This rule will check all cookies whose name ends with `mixpanel` and disable rule `921140`, if the value of any such cookie contains `%0a` (case-insensitive).

**Comment by fzipi** [comment]
Can we close this one?
